File: frontend/src/features/codexneo/schemas.ts
import { z } from "zod";

export const CodexNeoSettingsSchema = z.object({
  codexApiBaseUrl: z.string().trim().min(1),
  codexgoApiBaseUrl: z.string().trim().min(1),
  codexgoAutoRefreshEnabled: z.boolean(),
  codexgoAutoRefreshIntervalMinutes: z.number().int().min(5).max(1440),
  openaiActivityLogEnabled: z.boolean(),
  managementActivityLogEnabled: z.boolean(),
  codexHomeAutoRefreshEnabled: z.boolean(),
  codexHomeAutoRefreshIntervalSeconds: z.number().int().min(5).max(3600),
  codexHomeAutoSyncEnabled: z.boolean(),
  minimizeToTrayEnabled: z.boolean(),
  buyerTokenSaved: z.boolean(),
});

export const CodexNeoSettingsUpdateRequestSchema = z.object({
  codexApiBaseUrl: z.string().trim().min(1).optional(),
  codexgoApiBaseUrl: z.string().trim().min(1).optional(),
  codexgoAutoRefreshEnabled: z.boolean().optional(),
  codexgoAutoRefreshIntervalMinutes: z.number().int().min(5).max(1440).optional(),
  openaiActivityLogEnabled: z.boolean().optional(),
  managementActivityLogEnabled: z.boolean().optional(),
  codexHomeAutoRefreshEnabled: z.boolean().optional(),
  codexHomeAutoRefreshIntervalSeconds: z.number().int().min(5).max(3600).optional(),
  codexHomeAutoSyncEnabled: z.boolean().optional(),
  minimizeToTrayEnabled: z.boolean().optional(),
  buyerToken: z.string().optional(),
  clearBuyerToken: z.boolean().optional(),
});

export const CodexNeoApiUrlRequestSchema = z.object({
  codexApiBaseUrl: z.string().trim().min(1).optional(),
});

export const CodexNeoActionResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  configPath: z.string().nullable().optional(),
  authPath: z.string().nullable().optional(),
  backupPath: z.string().nullable().optional(),
  restartAttempted: z.boolean().optional(),
  restartSucceeded: z.boolean().nullable().optional(),
  restartOutput: z.string().nullable().optional(),
});

export const CodexNeoPathResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  path: z.string().nullable().optional(),
});

export const CodexNeoCodexHomeResponseSchema = z.object({
  codexHome: z.string(),
  defaultCodexHome: z.string(),
  dataDir: z.string(),
  customCodexHome: z.boolean(),
  exists: z.boolean(),
});

export const CodexNeoCodexHomeUpdateRequestSchema = z.object({
  codexHome: z.string().trim().min(1),
});

export const CodexNeoPathRequestSchema = z.object({
  path: z.string().optional(),
});

export const CodexNeoAccountKeysRequestSchema = z.object({
  accountKeys: z.array(z.string().trim().min(1)).min(1),
});

export const CodexNeoLocationRequestSchema = z.object({
  accountKeys: z.array(z.string().trim().min(1)).min(1),
  location: z.enum(["codex", "backup"]),
  present: z.boolean(),
});

export const CodexNeoBulkLocationRequestSchema = z.object({
  location: z.enum(["codex", "backup"]),
  present: z.boolean(),
});

export const CodexNeoValidityDateRequestSchema = z.object({
  accountKeys: z.array(z.string().trim().min(1)).min(1),
  validityDate: z.string().trim().min(1),
});

export const CodexNeoSwitchRequestSchema = z.object({
  accountKey: z.string().trim().min(1),
  restart: z.boolean(),
});

export const CodexNeoActivityLogResponseSchema = z.object({
  contents: z.string(),
});

export const CodexNeoHealthItemSchema = z.object({
  key: z.string(),
  label: z.string(),
  status: z.enum(["ok", "warning", "error"]),
  message: z.string(),
  detail: z.string().nullable().optional(),
  copyValue: z.string().nullable().optional(),
});

export const CodexNeoHealthResponseSchema = z.object({
  overallStatus: z.enum(["ok", "warning", "error"]),
  items: z.array(CodexNeoHealthItemSchema),
});

export const CodexNeoAccountUsageWindowSchema = z.object({
  usedPercent: z.number().int().nullable().optional(),
  resetsAt: z.string().nullable().optional(),
  windowMinutes: z.number().int().nullable().optional(),
});

export const CodexNeoAccountUsageSchema = z.object({
  primary: CodexNeoAccountUsageWindowSchema.nullable().optional(),
  secondary: CodexNeoAccountUsageWindowSchema.nullable().optional(),
});

export const CodexNeoAccountRowSchema = z.object({
  accountKey: z.string(),
  selector: z.string().nullable().optional(),
  email: z.string().nullable().optional(),
  alias: z.string().nullable().optional(),
  accountName: z.string().nullable().optional(),
  plan: z.string().nullable().optional(),
  authMode: z.string().nullable().optional(),
  active: z.boolean(),
  lastUsageAt: z.string().nullable().optional(),
  usage: CodexNeoAccountUsageSchema,
  codex: z.boolean(),
  backup: z.boolean(),
  api: z.boolean(),
  apiHourCount: z.number().int(),
  apiDayCount: z.number().int(),
  availability: z.string().nullable().optional(),
  status: z.string().nullable().optional(),
});

export const CodexNeoAccountsResponseSchema = z.object({
  accounts: z.array(CodexNeoAccountRowSchema),
  activeAccountKey: z.string().nullable().optional(),
  registryPath: z.string().nullable().optional(),
  message: z.string(),
  codexAllEnabled: z.boolean().optional(),
  backupAllEnabled: z.boolean().optional(),
});

export type CodexNeoSettings = z.infer<typeof CodexNeoSettingsSchema>;
export type CodexNeoSettingsUpdateRequest = z.infer<typeof CodexNeoSettingsUpdateRequestSchema>;
export type CodexNeoApiUrlRequest = z.infer<typeof CodexNeoApiUrlRequestSchema>;
export type CodexNeoLocationRequest = z.infer<typeof CodexNeoLocationRequestSchema>;
export type CodexNeoBulkLocationRequest = z.infer<typeof CodexNeoBulkLocationRequestSchema>;
export type CodexNeoActionResponse = z.infer<typeof CodexNeoActionResponseSchema>;
export type CodexNeoPathResponse = z.infer<typeof CodexNeoPathResponseSchema>;
export type CodexNeoCodexHomeResponse = z.infer<typeof CodexNeoCodexHomeResponseSchema>;
export type CodexNeoActivityLogResponse = z.infer<typeof CodexNeoActivityLogResponseSchema>;
export type CodexNeoHealthItem = z.infer<typeof CodexNeoHealthItemSchema>;
export type CodexNeoHealthResponse = z.infer<typeof CodexNeoHealthResponseSchema>;
export type CodexNeoAccountRow = z.infer<typeof CodexNeoAccountRowSchema>;
export type CodexNeoAccountsResponse = z.infer<typeof CodexNeoAccountsResponseSchema>;

File: app/modules/codexneo/schemas.py
from __future__ import annotations

from typing import Literal

from app.modules.shared.schemas import DashboardModel


class CodexNeoSettingsResponse(DashboardModel):
    codex_api_base_url: str
    codexgo_api_base_url: str
    codexgo_auto_refresh_enabled: bool
    codexgo_auto_refresh_interval_minutes: int
    openai_activity_log_enabled: bool
    management_activity_log_enabled: bool
    codex_home_auto_refresh_enabled: bool
    codex_home_auto_refresh_interval_seconds: int
    codex_home_auto_sync_enabled: bool
    minimize_to_tray_enabled: bool
    start_with_windows_enabled: bool
    auto_delete_free_reauth_accounts_enabled: bool
    auto_delete_quota_exceeded_accounts_enabled: bool
    buyer_token_saved: bool


class CodexNeoSettingsUpdateRequest(DashboardModel):
    codex_api_base_url: str | None = None
    codexgo_api_base_url: str | None = None
    codexgo_auto_refresh_enabled: bool | None = None
    codexgo_auto_refresh_interval_minutes: int | None = None
    openai_activity_log_enabled: bool | None = None
    management_activity_log_enabled: bool | None = None
    codex_home_auto_refresh_enabled: bool | None = None
    codex_home_auto_refresh_interval_seconds: int | None = None
    codex_home_auto_sync_enabled: bool | None = None
    minimize_to_tray_enabled: bool | None = None
    start_with_windows_enabled: bool | None = None
    auto_delete_free_reauth_accounts_enabled: bool | None = None
    auto_delete_quota_exceeded_accounts_enabled: bool | None = None
    buyer_token: str | None = None
    clear_buyer_token: bool = False


class CodexNeoApiUrlRequest(DashboardModel):
    codex_api_base_url: str | None = None


class CodexNeoActionResponse(DashboardModel):
    success: bool
    message: str
    config_path: str | None = None
    auth_path: str | None = None
    backup_path: str | None = None
    restart_attempted: bool = False
    restart_succeeded: bool | None = None
    restart_output: str | None = None


class CodexNeoPathResponse(DashboardModel):
    success: bool
    message: str
    path: str | None = None


class CodexNeoCodexHomeResponse(DashboardModel):
    codex_home: str
    default_codex_home: str
    data_dir: str
    custom_codex_home: bool
    exists: bool


class CodexNeoCodexHomeUpdateRequest(DashboardModel):
    codex_home: str


class CodexNeoPathRequest(DashboardModel):
    path: str | None = None


class CodexNeoAccountKeysRequest(DashboardModel):
    account_keys: list[str]


class CodexNeoLocationRequest(DashboardModel):
    account_keys: list[str]
    location: Literal["codex", "backup"]
    present: bool


class CodexNeoBulkLocationRequest(DashboardModel):
    location: Literal["codex", "backup"]
    present: bool


class CodexNeoValidityDateRequest(DashboardModel):
    account_keys: list[str]
    validity_date: str


class CodexNeoSwitchRequest(DashboardModel):
    account_key: str
    restart: bool = False


class CodexNeoActivityLogResponse(DashboardModel):
    contents: str


class CodexNeoHealthItem(DashboardModel):
    key: str
    label: str
    status: str
    message: str
    detail: str | None = None
    copy_value: str | None = None


class CodexNeoHealthResponse(DashboardModel):
    overall_status: str
    items: list[CodexNeoHealthItem]


class CodexNeoAccountUsageWindow(DashboardModel):
    used_percent: int | None = None
    remaining_percent: int | None = None
    resets_at: str | None = None
    window_minutes: int | None = None


class CodexNeoAccountUsage(DashboardModel):
    primary: CodexNeoAccountUsageWindow | None = None
    secondary: CodexNeoAccountUsageWindow | None = None


class CodexNeoAccountRow(DashboardModel):
    account_key: str
    selector: str | None = None
    email: str | None = None
    alias: str | None = None
    account_name: str | None = None
    plan: str | None = None
    auth_mode: str | None = None
    active: bool = False
    last_usage_at: str | None = None
    usage: CodexNeoAccountUsage = CodexNeoAccountUsage()
    codex: bool = False
    backup: bool = False
    api: bool = False
    api_hour_count: int = 0
    api_day_count: int = 0
    availability: str | None = None
    status: str | None = None
    codex_ib_account_id: str | None = None
    codex_ib_status: str | None = None
    codex_ib_status_reason: str | None = None
    codex_ib_routable: bool | None = None


class CodexNeoAccountsResponse(DashboardModel):
    accounts: list[CodexNeoAccountRow]
    active_account_key: str | None = None
    registry_path: str | None = None
    message: str
    codex_all_enabled: bool = False
    backup_all_enabled: bool = False
